### ISSUE TITLE
Break dependency loop by moving `What4.TransitionSystem.Exporter.Sally` to `language-sally` 

### ISSUE THREAD

**Issue by RyanGlScott** [GaloisInc/what4#186]
_(Migrated from the discussion starting at https://github.com/GaloisInc/what4/pull/131#issuecomment-860863901.)_

Currently, the `what4-transition-system` library depends on the [`language-sally`](https://github.com/GaloisInc/language-sally) library. As a result, the `what4` repo checks in the `language-sally` repo as a submodule. However, the `language-sally` library depends on the `what4` library. This poses a maintenance challenge. Things like making version bumps require coordinated changes between the `what4` and `language-sally` repos to get right. Moreover, changes made to `what4` can cause `language-sally` to break without maintainers noticing for a while. For instance, `language-sally` currently fails to build against commit ac64bcd580f552bd22ec5a135fdbcc3d523723a1 of the `what4` repo:

```
[2 of 5] Compiling Language.Sally.Writer ( src/Language/Sally/Writer.hs, interpreted )

src/Language/Sally/Writer.hs:204:23: error:
    Not in scope: ‘SMT2.forall’
    Perhaps you meant ‘SMT2.forall_’ (imported from What4.Protocol.SMTLib2.Syntax)
    Module ‘What4.Protocol.SMTLib2.Syntax’ does not export ‘forall’.
    |
204 |   forallExpr vars t = SMT2.forall (varBinding @a <$> vars) <$> t
    |                       ^^^^^^^^^^^

src/Language/Sally/Writer.hs:205:23: error:
    Not in scope: ‘SMT2.exists’
    Perhaps you meant one of these:
      ‘SMT2.exists_’ (imported from What4.Protocol.SMTLib2.Syntax),
      ‘SMT2.exit’ (imported from What4.Protocol.SMTLib2.Syntax)
    Module ‘What4.Protocol.SMTLib2.Syntax’ does not export ‘exists’.
    |
205 |   existsExpr vars t = SMT2.exists (varBinding @a <$> vars) <$> t
    |                       ^^^^^^^^^^^
```

We should break this dependency loop by moving the `What4.TransitionSystem.Exporter.Sally` module from `what4-transition-system` to the `language-sally` repo. Once this is done, `what4-transition-system` will no longer need to depend on `language-sally`, allowing the `what4` repo to remove `language-sally` as a submodule.

cc @Ptival

**Comment by Ptival** [comment]
Agreed! I'm very busy until tomorrow, but can try and get this sorted some time after!

**Comment by RyanGlScott** [comment]
Thanks, @Ptival! And I'm not in a particular hurry to get this done—I just noticed that this complicates the story of trying to update What4 to support GHC 9.2.

**Comment by RyanGlScott** [comment]
This was done in GaloisInc/what4#190 and GaloisInc/language-sally#8.
